internal/cephfs/controllerserver.go

```diff
@@ -59,52 +60,83 @@ func (cs *ControllerServer) createBackingVolume(
 	volOptions,
 	parentVolOpt *store.VolumeOptions,
 	pvID *store.VolumeIdentifier,
-	sID *store.SnapshotIdentifier) error {
+	sID *store.SnapshotIdentifier,
+) error {
 	var err error
 	volClient := core.NewSubVolume(volOptions.GetConnection(), &volOptions.SubVolume, volOptions.ClusterID)
 
 	if sID != nil {
-		if err = cs.OperationLocks.GetRestoreLock(sID.SnapshotID); err != nil {
-			log.ErrorLog(ctx, err.Error())
+		return cs.createBackingVolumeFromSnapshotSource(ctx, volOptions, parentVolOpt, volClient, sID)
+	}
 
-			return status.Error(codes.Aborted, err.Error())
-		}
-		defer cs.OperationLocks.ReleaseRestoreLock(sID.SnapshotID)
-		snap := core.Snapshot{
-			SnapshotID: sID.FsSnapshotName,
-			SubVolume:  &parentVolOpt.SubVolume,
-		}
+	if parentVolOpt != nil {
+		return cs.createBackingVolumeFromVolumeSource(ctx, parentVolOpt, volClient, pvID)
+	}
 
-		err = volClient.CreateCloneFromSnapshot(ctx, snap)
-		if err != nil {
-			log.ErrorLog(ctx, "failed to create clone from snapshot %s: %v", sID.FsSnapshotName, err)
+	if err = volClient.CreateVolume(ctx); err != nil {
+		log.ErrorLog(ctx, "failed to create volume %s: %v", volOptions.RequestName, err)
 
-			return err
-		}
+		return status.Error(codes.Internal, err.Error())
+	}
 
-		return err
+	return nil
+}
+
+func (cs *ControllerServer) createBackingVolumeFromSnapshotSource(
+	ctx context.Context,
+	volOptions *store.VolumeOptions,
+	parentVolOpt *store.VolumeOptions,
+	volClient core.SubVolumeClient,
+	sID *store.SnapshotIdentifier,
+) error {
+	if err := cs.OperationLocks.GetRestoreLock(sID.SnapshotID); err != nil {
+		log.ErrorLog(ctx, err.Error())
+
+		return status.Error(codes.Aborted, err.Error())
 	}
-	if parentVolOpt != nil {
-		if err = cs.OperationLocks.GetCloneLock(pvID.VolumeID); err != nil {
-			log.ErrorLog(ctx, err.Error())
+	defer cs.OperationLocks.ReleaseRestoreLock(sID.SnapshotID)
 
-			return status.Error(codes.Aborted, err.Error())
-		}
-		defer cs.OperationLocks.ReleaseCloneLock(pvID.VolumeID)
-		err = volClient.CreateCloneFromSubvolume(
-			ctx, &parentVolOpt.SubVolume)
-		if err != nil {
-			log.ErrorLog(ctx, "failed to create clone from subvolume %s: %v", fsutil.VolumeID(pvID.FsSubvolName), err)
+	if volOptions.BackingSnapshot {
+		if err := store.AddSnapshotBackedVolumeRef(ctx, volOptions); err != nil {
+			log.ErrorLog(ctx, "failed to create snapshot-backed volume from snapshot %s: %v",
+				sID.FsSnapshotName, err)
 
 			return err
 		}
 
 		return nil
 	}
-	if err = volClient.CreateVolume(ctx); err != nil {
-		log.ErrorLog(ctx, "failed to create volume %s: %v", volOptions.RequestName, err)
 
-		return status.Error(codes.Internal, err.Error())
+	err := volClient.CreateCloneFromSnapshot(ctx, core.Snapshot{
```

Comment:
if clone creation fails we need to remove the VolRef added?

Reply:
@Madhu-1 I guess you mean `store.AddSnapshotBackedVolumeRef()`? If this call succeeds, the function returns early (it's all in `volOptions.BackingSnapshot` condition). Because that's all that is needed for snapshot-backed volumes -- only reserving a volume ID and adding a ref to the reference tracker for the snapshot. 

https://github.com/ceph/ceph-csi/blob/4fa98d40588ef380ce0de46da8bf30ca6673f47c/internal/cephfs/controllerserver.go#L99-L118

So I think this should be ok? Please let me know if I missed something.

Reply:
what I meant was If this CreateCloneFromSnapshot fails we need to do undo operation of AddSnapshotBackedVolumeRef to avoid leaks?

```go
if volOptions.BackingSnapshot {
		if err := store.AddSnapshotBackedVolumeRef(ctx, volOptions); err != nil {
			log.ErrorLog(ctx, "failed to create snapshot-backed volume from snapshot %s: %v",
				sID.FsSnapshotName, err)

			return err
		}

		return nil
	}

	err := volClient.CreateCloneFromSnapshot(ctx, core.Snapshot{ // If this calls fails we need to do undo operation of AddSnapshotBackedVolumeRef to avoid leaks?
		SnapshotID: sID.FsSnapshotName,
		SubVolume:  &parentVolOpt.SubVolume,
	})
```

Reply:
@Madhu-1 but if `volOptions.BackingSnapshot` is `true`, the function never reaches `volClient.CreateCloneFromSnapshot()`. I might be missing something obvious, but can you please check again?

Reply:
Yes, correct I missed that `If`  check :+1: this looks good. Thanks for checking this one.